### PR TITLE
Disk removal control and recovery

### PR DESCRIPTION
**Pull request by Justarone**

### CODE REVIEW

File: src/core/backend/pearl/disk_controller.rs [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+            check_interval.tick().await;
+        }
+    }
+
+    async fn monitor_task(self: Arc<Self>) {
+        let mut check_interval = interval(CHECK_INTERVAL);
+        Self::monitor_wait(self.state.clone(), &mut check_interval).await;
+        loop {
+            check_interval.tick().await;
+            if !Self::is_work_dir_available(self.disk.path()) {
```
Comment (pyakushin) [inline src/core/backend/pearl/disk_controller.rs:74]: Unnecessary boolean `not` operation

File: src/core/backend/pearl/disk_controller.rs [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    }
+
+    pub(crate) async fn init(&self) -> BackendResult<()> {
+        if *self.state.read().await == GroupsState::NotReady {
+            let groups = self.get_groups()?;
+            *self.groups.write().await = groups;
+            *self.state.write().await = GroupsState::Initialized;
+            Ok(())
+        } else {
+            Err(Error::internal().into())
```
Comment (pyakushin) [inline src/core/backend/pearl/disk_controller.rs:125]: Conversion to the same type

File: src/core/backend/pearl/disk_controller.rs [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+                let dump_sem = Arc::new(Semaphore::new(self.init_par_degree));
+                let path = self.settings.normal_path(self.disk.path(), vdisk_id);
+                Group::new(
+                    self.settings.clone(),
+                    vdisk_id,
+                    self.node_name.clone(),
+                    self.disk.name().to_owned(),
+                    path,
+                    self.node_name.clone(),
+                    dump_sem.clone(),
```
Comment (pyakushin) [inline src/core/backend/pearl/disk_controller.rs:151]: Redundant clone

File: src/core/backend/pearl/disk_controller.rs [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        Ok(())
+    }
+
+    pub(crate) async fn run(&self) -> Result<()> {
+        if *self.state.read().await == GroupsState::Initialized {
+            if let Err(e) = Self::run_groups(self.groups.clone()).await {
+                error!("Can't run groups on disk {:?} (reason: {})", self.disk, e);
+                // if work dir became unavailable we should reinit, so new state is NotReady
+                // otherwise - we'll try again
+                if !Self::is_work_dir_available(self.disk.path().clone()) {
```
Comment (pyakushin) [inline src/core/backend/pearl/disk_controller.rs:223]: Redundant clone, you call clone on the double-reference, btw

File: src/core/backend/pearl/disk_controller.rs [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+                    match storage.close().await {
+                        Ok(_) => debug!("holder {} closed", id),
+                        Err(e) => {
+                            error!("error closing holder{}: {} (disk: {:?})", id, e, self.disk)
+                        }
+                    }
+                });
+            }
+        }
+        let _ = futures.collect::<()>().await;
```
Comment (pyakushin) [inline src/core/backend/pearl/disk_controller.rs:382]: You assingning `()` unit value

**Review by pyakushin — APPROVED** [review]

File: src/core/backend/pearl/disk_controller.rs (hunk 5 above)
Comment (Justarone) [reply src/core/backend/pearl/disk_controller.rs:382]: This code was written before me (in Pearl strurcture), so I decided to leave it as it is (but seems that there is nothing special in this expression and you confirmed that so I remove it)

**Pull request merged**